llvm/tools/llvm-reduce/deltas/ReduceValuesToReturn.cpp

```diff
@@ -0,0 +1,244 @@
+//===----------------------------------------------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// Try to reduce a function by inserting new return instructions. Try to insert
+// an early return for each instruction value at that point. This requires
+// mutating the return type, or finding instructions with a compatible type.
+//
+//===----------------------------------------------------------------------===//
+
+#define DEBUG_TYPE "llvm-reduce"
+
+#include "ReduceValuesToReturn.h"
+
+#include "Delta.h"
+#include "Utils.h"
+#include "llvm/IR/CFG.h"
+#include "llvm/IR/Instructions.h"
+#include "llvm/Transforms/Utils/BasicBlockUtils.h"
+#include "llvm/Transforms/Utils/Cloning.h"
+
+using namespace llvm;
+
+/// Return true if it is legal to emit a copy of the function with a non-void
+/// return type.
+static bool canUseNonVoidReturnType(const Function &F) {
+  // Functions with sret arguments must return void.
+  return !F.hasStructRetAttr() &&
+         CallingConv::supportsNonVoidReturnType(F.getCallingConv());
+}
+
+/// Return true if it's legal to replace a function return type to use \p Ty.
+static bool isReallyValidReturnType(Type *Ty) {
+  return FunctionType::isValidReturnType(Ty) && !Ty->isTokenTy() &&
+         Ty->isFirstClassType();
+}
+
+/// Insert a ret inst after \p NewRetValue, which returns the value it produces.
+static void rewriteFuncWithReturnType(Function &OldF, Value *NewRetValue) {
+  Type *NewRetTy = NewRetValue->getType();
+  FunctionType *OldFuncTy = OldF.getFunctionType();
+
+  FunctionType *NewFuncTy =
+      FunctionType::get(NewRetTy, OldFuncTy->params(), OldFuncTy->isVarArg());
+
+  LLVMContext &Ctx = OldF.getContext();
+  Instruction *NewRetI = cast<Instruction>(NewRetValue);
+  BasicBlock *NewRetBlock = NewRetI->getParent();
+
+  BasicBlock::iterator NewValIt = NewRetI->getIterator();
+
+  // Hack up any return values in other blocks, we can't leave them as ret void.
+  if (OldFuncTy->getReturnType()->isVoidTy()) {
+    for (BasicBlock &OtherRetBB : OldF) {
+      if (&OtherRetBB != NewRetBlock) {
+        auto *OrigRI = dyn_cast<ReturnInst>(OtherRetBB.getTerminator());
+        if (!OrigRI)
+          continue;
+
+        OrigRI->eraseFromParent();
+        ReturnInst::Create(Ctx, getDefaultValue(NewRetTy), &OtherRetBB);
+      }
+    }
+  }
+
+  // Now prune any CFG edges we have to deal with.
+  //
+  // Use KeepOneInputPHIs in case the instruction we are using for the return is
+  // that phi.
+  // TODO: Could avoid this with fancier iterator management.
+  for (BasicBlock *Succ : successors(NewRetBlock))
+    Succ->removePredecessor(NewRetBlock, /*KeepOneInputPHIs=*/true);
```

Comment:
In terms of general approach, maybe it would be simpler to first call splitBasicBlock(), replace the terminator with return, and then leave everything else to EliminateUnreachableBlocks? I think that should also automatically handle the unusual terminators in a sensible way.

Reply:
Done in #134794